src/components/MessageInput/index.js

```diff
@@ -0,0 +1,144 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import styles from './styles.scss';
+import i18n from './i18n';
+
+const UIHeightOffset = 23;
+// the extra height of the entire field with paddings and borders
+
+export default class MessageInput extends Component {
+  static propTypes = {
+    value: PropTypes.string.isRequired,
+    currentLocale: PropTypes.string.isRequired,
+    disabled: PropTypes.bool,
+    minHeight: PropTypes.number,
+    maxHeight: PropTypes.number,
+    maxLength: PropTypes.number,
+    onSend: PropTypes.func,
+    onChange: PropTypes.func,
+    onHeightChange: PropTypes.func,
+  }
+  static defaultProps = {
+    disabled: false,
+    onSend: undefined,
+    onChange: undefined,
+    onHeightChange: undefined,
+    minHeight: 63,
+    maxHeight: 300,
+    maxLength: 5000,
+  }
+  constructor(props, context) {
+    super(props, context);
+    this.state = {
+      value: props.value,
+      height: props.minHeight,
+    };
+  }
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.value !== this.state.value) {
+      // use setState(updater, callback) to recaculate height after value has been update to DOM
+      this.setState(
+        () => ({
+          value: nextProps.value,
+        }),
+        () => {
+          const newHeight = this.calculateNewHeight();
+          if (newHeight !== this.state.height) {
+            if (typeof this.props.onHeightChange === 'function') {
+              this.props.onHeightChange(newHeight);
+            }
+            this.setState({
+              height: newHeight,
+            });
+          }
+        }
+      );
+    }
+  }
+  calculateNewHeight() {
+    // temperarily set height to 0 to check scrollHeight
+    this.textArea.style.height = 0;
+    const newHeight = this.textArea.scrollHeight + 10 + UIHeightOffset;
+    // set height back to original to avoid messing with react
+    this.textArea.style.height = `${this.state.height - UIHeightOffset}px`;
+    const {
+      minHeight,
+      maxHeight,
+    } = this.props;
+    if (newHeight < minHeight) {
+      return minHeight;
+    } else if (newHeight > maxHeight) {
+      return maxHeight;
+    }
+    return newHeight;
+  }
+  onChange = (e) => {
+    const {
+      currentTarget: {
+        value,
+      },
+    } = e;
+    const newHeight = this.calculateNewHeight();
+    if (newHeight !== this.state.height && typeof this.props.onHeightChange === 'function') {
+      this.props.onHeightChange(newHeight);
+    }
+    this.setState({
+      value,
+      height: newHeight,
+    });
+    if (typeof this.props.onChange === 'function') {
+      this.props.onChange(value);
+    }
+  }
+  onKeyDown = (e) => {
+    if (e.key === 'Enter') {
+      if (e.shiftKey) {
+        /* ignore */
+      } else {
+        e.preventDefault();
+        if (typeof this.props.onSend === 'function') {
+          this.props.onSend();
+        }
+      }
+    }
```

Comment:
any particular reason to write in this way other than `if (!e.shiftKey) {}`?

Reply:
Fixed. I thought there might be something to do if e.shiftKey is true, forgot to clean up the code afterwards.